File: scripts/toc.rb
#!/usr/bin/env ruby

# == Synopsis
# toc : Generate a table of contents for an html file and append it to the top of the file.
#
# == Usage
#
# toc [OPTIONS]
#
# -h, --help ::
#   show this help message
# -t, --tags ::
#   comma separated list of tags to put in the table of contents
#   e.g. h2,h3
# -i, --infile ::
#   (X)HTML file input to generate TOC for (default is stdin).
# -o, --outfile ::
#   file to send output to (default is stdout)
#
# == Author
#
# Srikanth K Agaram (code at srikanthak dot name)
#
# == Copyright
#
# Copyright (c) 2009 Srikanth Agaram. Licensed under GNU GPLv3

require 'rdoc/usage'
require 'getoptlong'

require 'rubygems'
require 'builder'
require 'hpricot'

# Tree implementation to store the table of contents {{{
class TreeNode
  attr_accessor :parent, :depth, :data, :children

  def initialize(data)
    @data = data
    @parent = nil
    @children = []
  end

  def append(child)
    child.parent = self
    children.push child
  end

  def prefix_traverse(depth, &block)
    yield self
    children.each { |child| child.prefix_traverse(depth + 1, &block) }
  end

  def index
    @parent.children.index(self) + 1
  end

  def position
    if @parent.nil?
      []
    else
      @parent.position.push index
    end
  end

  def path
    position.join('_')
  end

  def depth
    if @parent.nil?
      0
    else
      @parent.depth + 1
    end
  end
end

class Tree
  attr_accessor :root

  def initialize(node)
    @root = node
  end

  def prefix_traverse(&block)
    @root.prefix_traverse(0, &block)
  end
end

# Example of building and traversing a tree {{{
# root = TreeNode.new('root')
# a = TreeNode.new('a')
# b = TreeNode.new('b')
# c = TreeNode.new('c')
# d = TreeNode.new('d')
# e = TreeNode.new('e')
# f = TreeNode.new('f')
# g = TreeNode.new('g')
# h = TreeNode.new('h')
# i = TreeNode.new('i')
# j = TreeNode.new('j')
# k = TreeNode.new('k')

# root.append(a)
# root.append(b)
# root.append(c)
# a.append(d)
# a.append(e)
# b.append(f)
# c.append(g)
# c.append(h)
# d.append(i)
# d.append(j)
# d.append(k)

# tree = Tree.new(root)

# tree.prefix_traverse do |depth, data|
#   depth.times do
#     print '.'
#   end
#   puts data
# end
# }}}

# }}}

# Table of contents extractor {{{
class Toc
  def initialize(options)
    default_options = {
      :tags => ['h2', 'h3'],
    }

    options = default_options.merge(options)

    if options[:infile].nil?
      @document = Hpricot $stdin
    else
      @document = Hpricot open(options[:infile])
    end

    @tags_list = options[:tags]
    @parent = gen_parent_hash

    @prefix = 'section_'

    output = gen_html
    if options[:outfile].nil?
      puts output
    else
      File.open(options[:outfile], 'w') do |file|
        file.write html
      end
    end
  end

  private

  # Helper functions {{{
  def gen_html
    # remove the old TOC if present
    @document.search('#toc-toc').remove

    # generate the TOC tree
    toc_tree = gen_toc_tree

    # add anchors to the TOC tree nodes
    add_toc_anchors(toc_tree)

    # generate the TOC html
    x = Builder::XmlMarkup.new(:indent => 2)
    x.div(:id => 'toc-toc') do
      x.h2 "Table of Contents"
      gen_toc_html(x, toc_tree.root)
    end

    # insert the TOC into the html
    @document.search('body').prepend(x.target!)

    # return the modified html
    @document.to_html
  end

  def gen_toc_tree
    elements = find_tags
    current_node = {}
    @tags_list.each { |tag| current_node[tag] = nil }

    toc_tree = Tree.new(TreeNode.new('root'))
    elements.each do |element|
      name = element.name.downcase
      new_node = TreeNode.new(element)
      current_node[name] = new_node

      if @parent[name].nil?
        toc_tree.root.append(new_node)
      else
        current_node[@parent[name]].append(new_node)
      end
    end

    toc_tree
  end

  def find_tags
    elements = @document.search("body//*").select do |e|
      result = false
      if e.methods.include? "name"
        result = @tags_list.include? e.name
      end
      result
    end

    elements
  end

  def add_toc_anchors(toc_tree)
    toc_tree.prefix_traverse do |node|
      unless node.parent.nil?
        # save the contents of the element
        html = node.data.inner_html

        # create the anchor element that will be linked to in the TOC
        anchor = Builder::XmlMarkup.new(:indent => 2)
        anchor.a(html, :name => @prefix + node.path)

        # insert the anchor into the element
        node.data.inner_html = anchor.target!
      end
    end
  end

  def gen_toc_html(builder, node)
    unless node.children.empty?
      builder.ul do
        node.children.each do |child|
          builder.li do
            builder.a(child.data.inner_text, :href => "##{@prefix}#{child.path}")
            gen_toc_html(builder, child)
          end
        end
      end
    end
  end

  def gen_parent_hash
    current = nil
    parent = {}
    @tags_list.each do |tag|
      parent[tag] = current unless current.nil?
      current = tag
    end

    parent
  end
  # }}}

end
# }}}

# Extract command line options {{{
options = {}

opts = GetoptLong.new(
        ['--help', '-h', GetoptLong::OPTIONAL_ARGUMENT],
        ['--tags', '-t', GetoptLong::OPTIONAL_ARGUMENT],
        ['--infile', '-i', GetoptLong::OPTIONAL_ARGUMENT],
        ['--outfile', '-o', GetoptLong::OPTIONAL_ARGUMENT]
                     )

opts.each do |opt, arg|
  case opt
  when '--help'
    case arg
    when /^u/i
      RDoc::usage('Usage')
    when /^a/i
      RDoc::usage('Author')
    when /l|c/i
      RDoc::usage('Copyright')
    else
      RDoc::usage
    end

  when '--tags'
    options[:tags] = arg.split(',')

  when '--infile'
    options[:infile] = arg

  when '--outfile'
    options[:outile] = arg
  end
end
# }}}

toc = Toc.new(options)

File: scripts/toc.py
#!/usr/bin/env python
import sys
import getopt
from BeautifulSoup import BeautifulSoup, Tag, NavigableString

class Toc:

  def __init__(self, title, options):
    # defaults
    self.infile = sys.stdin
    self.outfile = sys.stdout
    self.tag_names = ['h2', 'h3']
    self.toc_id = 'auto_toc'
    self.name_prefix = 'section'
    self.title = title

    # options
    for option, value in options:
      if option in ('-h', '--help'):
        usage()
        sys.exit()
      elif option in ('-t', '--tags'):
        self.tag_names = value.split(',')
      elif option in ('-i', '--infile'):
        self.infile = open(value, 'r')
      elif option in ('-o', '--outfile'):
        self.outfile = open(value, 'w')

    # process the html, create toc and print
    self.get_tags()
    if self.tag_list:
      self.id_tags()
      self.create_toc()
    self.output()

    # clean up
    self.infile.close()
    self.outfile.close()

  def get_tags(self):
    self.soup = BeautifulSoup(self.infile.read())

    # check if there is an existing toc
    toc = self.soup.findAll(id=self.toc_id)
    for tag in toc:
      tag.extract()

    # check which of the mentioned tags are present
    tag_names = []
    for tag_name in self.tag_names:
      tag_list = self.soup.findAll(tag_name)
      if tag_list:
        tag_names.append(tag_name)
      if len(tag_names) >= 2:
        break
    self.tag_names = tag_names

    # get tags
    self.tag_list = self.soup.findAll(self.tag_names) if self.tag_names else []

  def id_tags(self):
    counts = []
    self.toc_list = []
    for item in self.tag_names:
      counts.append(0)
    for tag in self.tag_list:
      reset = False
      depth = 0
      for index, tag_name in enumerate(self.tag_names):
        if reset == True:
          counts[index] = 0
        if tag.name == tag_name:
          depth = index
          counts[index] += 1
          reset = True

      name = self.name_prefix
      for count in counts:
        if count == 0:
          break
        name = '%(name)s_%(count)i' % { 'name': name, 'count': count }

      tag['id'] = name
      self.toc_list.append({ 'depth': depth, 'id': name, 'title': tag.text })

  def create_toc(self):
    # lists will hold the last ol/ul elements at each depth
    lists = []

    # setup the toc container
    toc = Tag(self.soup, 'div')
    toc['id'] = self.toc_id
    last_li = toc
    header = Tag(self.soup, 'h2')
    header_title = NavigableString('Contents')
    header.append(header_title)
    toc.append(header)

    for toc_item in self.toc_list:
      depth = toc_item['depth']
      if len(lists) ==  depth: # this is the first time we're at this depth
        list_el = Tag(self.soup, 'ol')
        lists.append(list_el)
        last_li.append(list_el)

      elif depth > old_depth: # this is a new sub-tree
        list_el = Tag(self.soup, 'ol')
        lists[depth] = list_el
        last_li.append(list_el)

      old_depth = depth

      # set up the new item
      li = Tag(self.soup, 'li')
      a = Tag(self.soup, 'a')
      title = NavigableString(toc_item['title'])
      a.append(title)
      a['href'] = '#%(id)s' % { 'id': toc_item['id'] }
      li.append(a)
      lists[depth].append(li)
      last_li = li

    # insert the toc at the top of the html
    self.soup.insert(0, toc)

  def output(self):
    prefix = """<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
	<title>%(title)s</title>
</head>
<body>
    """ % { 'title': self.title }
    suffix = """
    </body>
    </html>
    """
    self.outfile.write(prefix)
    self.outfile.write(self.soup.prettify())
    self.outfile.write(suffix)

def parse_opts(args):
  try:
    opts, args = getopt.getopt(args, 'ht:i:o:', ['help', 'tags', 'infile', 'outfile'])
  except getopt.GetoptError, err:
    sys.stderr.write(str(err))
    usage()
    sys.exit(2)
  return ' '.join(args), opts

def usage():
  print
  print 'usage: %(name)s [options]' % { 'name': sys.argv[0] }
  print 'options:'
  print '\t-h, --help ::'
  print '\t\tshow this help message'
  print '\t-t, --tags ::'
  print '\t\tcomma separated list of tags to put in the table of contents'
  print '\t\te.g. h2,h3'
  print '\t-i, --infile ::'
  print '\t\t(X)HTML file input to generate TOC for (default is stdin).'
  print '\t-o, --outfile ::'
  print '\t\tfile to send output to (default is stdout)'

if __name__ == '__main__':
  title, options = parse_opts(sys.argv[1:])
  toc = Toc(title, options)
